File: AOC_2023/Day 10/10.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

type Point struct {
	x int
	y int
}

func initialize(filename string) ([]string, error) {
	file, err := os.Open(filename)
	if err != nil {
		return nil, err
	}
	defer file.Close()
	var lines []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}
	if err := scanner.Err(); err != nil {
		return nil, err
	}
	return lines, nil
}

func part1(input []string) [] Point{
	pipes := map[rune][]rune{
		'|' : {'N', 'S'},
		'-' : {'E', 'W'},
		'L' : {'N', 'E'},
		'J' : {'N', 'W'},
		'7' : {'S', 'W'},
		'F' : {'S', 'E'},
		'S' : {'N','S', 'E', 'W'},
	}
	directions := map[rune]Point{
		'N' : {-1, 0},
		'S' : {1, 0},
		'E' : {0, 1},
		'W' : {0, -1},
	}
	directions_to_pipes := map[rune]string{
		'N' : "|7FS",
		'S' : "|JLS",
		'E' : "-7JS",
		'W' : "-LFS",
	}
	parent := map[Point]Point{}
	var path []Point
	var start Point
	for i, line := range input {
		for j, char := range line {
			if char == 'S' {
				start = Point{i, j}
				break
			}
		}
	}
	queue := make([]Point, 0)
	queue = append(queue, start)
	non := Point{-1, -1}
	parent[start] = non
	for len(queue) > 0 {
		curr := queue[0]
		queue = queue[1:]
		if curr == start && parent[start] != non {break}
		if input[curr.x][curr.y] == '.' {continue}
		for _, dir := range pipes[rune(input[curr.x][curr.y])] {
			x := curr.x + directions[dir].x
			y := curr.y + directions[dir].y
			if x >= 0 && x < len(input) && y >= 0 && y < len(input[0])  && strings.Contains(directions_to_pipes[dir], string(input[x][y])) {
				neighbour := Point{x, y}
				if parent[curr] == neighbour {continue}
				parent[neighbour] = curr
				queue = append(queue, neighbour)
			}
		}
	}
	path = append(path, start)
	for true{
		curr := parent[path[len(path)-1]]
		if curr == start {break}
		path = append(path, curr)
	}
	fmt.Println("Part 1:", len(path)/2)
	return path
}

func part2(path []Point) {
	path = append(path, path[0])
	area := 0
	for i := 0; i < len(path)-1; i++ {
		area += path[i].x * path[i+1].y - path[i].y * path[i+1].x
	}
	result := area/2
	if result < 0 {result *= -1}
	result -= len(path)/2
	result += 1
	fmt.Println("Part 2:", result)
}

func main(){
	filename := "input.txt"
	input, err := initialize(filename)
	if err != nil {
		fmt.Println("Error reading file:", err)
		return
	}
	path := part1(input)
	part2(path)
}

File: AOC_2023/Day 10/10.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <algorithm>
#include <queue>
#include <map>

using namespace std;

class Solution
{
    public:
    int n, m;
    vector<string> grid;
    map<char, vector<char>> pipes = {
        {'|', {'N', 'S'}},
        {'-', {'E', 'W'}},
        {'L', {'N', 'E'}},
        {'J', {'N', 'W'}},
        {'7', {'S', 'W'}},
        {'F', {'S', 'E'}},
        {'S', {'S', 'E', 'W', 'N'}},
    };
    map<char,pair<int,int>> directions = {
        {'N', {-1, 0}},
        {'S', {1, 0}},
        {'E', {0, 1}},
        {'W', {0, -1}},
    };
    map<char, string> directions_to_pipe = {
        {'N', "|7FS"},
        {'S', "|JLS"},
        {'E', "-J7S"},
        {'W', "-LFS"},
    };
    map<pair<int,int>,pair<int,int>> parent;
    vector<pair<int,int>> path;
    void init(ifstream& input){
        string line;
        while (getline(input, line))
        {
            grid.push_back(line);
        }
        n = grid.size();
        m = grid[0].size();
    }
    bool withinBounds(int i, int j){
        return i >= 0 && i < n && j >= 0 && j < m;
    }
    vector<pair<int,int>> getNeighbours(int i, int j){
        vector<pair<int,int>> neighbours;
        if(grid[i][j] == '.') return neighbours;
        for(auto dir : pipes[grid[i][j]]){
            int x = directions[dir].first, y = directions[dir].second;
            if(withinBounds(i+x, j+y) && directions_to_pipe[dir].find(grid[i+x][j+y]) != string::npos){
                neighbours.push_back({i+x, j+y});
            }
        }
        return neighbours;
    }
    void part1(){
        queue<pair<int, int>> q;
        // find the starting point
        pair<int,int> start;
        for(int i = 0; i < grid.size(); i++){
            for(int j = 0; j < grid[0].size(); j++){
                if(grid[i][j] == 'S'){
                    start = {i, j};
                    q.push(start);
                    break;
                }
            }
        }
        parent[start] = {-1, -1};
        while(!q.empty()){
            auto curr = q.front();
            q.pop();
            int i = curr.first, j = curr.second;
            if(curr == start && parent[start] != make_pair(-1, -1)) break;
            for (auto neighbour : getNeighbours(i, j))
            {
                if(parent[curr] == neighbour) continue;
                parent[neighbour] = curr;
                q.push(neighbour);
            }
        }
        path.push_back(start);
        while(true){
            auto curr = parent[path.back()];
            if(curr == start) break;
            path.push_back(curr);
        }
        cout<<"Part 1: "<<path.size()/2<<"\n";
    }
    
    // Pick's theorem for area of polygon
    // A = i + b/2 - 1
    // i = number of points inside the polygon = A - b/2 + 1
    // We can find the area of the polygon by shoelace formula
    // A = 1/2 * sum(x[i]*y[i+1] - x[i+1]*y[i])
    // b = number of points on the boundary of the polygon
    void part2(){
        path.push_back(path[0]);
        int area = 0;
        for(int i = 0; i < path.size()-1; i ++){
            area += path[i].first*path[i+1].second - path[i+1].first*path[i].second;
        }
        if(area < 0) area *= -1;
        area /= 2;
        int result = area - path.size()/2 + 1;
        cout<<"Part 2: "<<result<<"\n";
    }   

};

int main(){
    ifstream in("input.txt");
    if (!in.is_open())
    {
        cerr << "Could not open the file - '" << endl;
        return EXIT_FAILURE;
    }
    Solution s;
    s.init(in);
    s.part1();
    s.part2();
    return 0;
}